blur/rand.py

```diff
@@ -413,42 +413,43 @@ def weighted_shuffle(weights):
     Raises:
         ValueError: If passed ``weights`` is not formed correctly
     """
-    working_list = weights[:]
-    # list of tuples [list_item, target_index]
-    # used to store positions before moving them into the final list
-    shuffle_positions = []
-
-    def closest_available(requested_index):
-        """The closest index that isn't taken in ``shuffle_positions``."""
-        taken_positions = [position[1] for position in shuffle_positions]
-        available_indexes = [i for i in range(len(weights))
-                             if i not in taken_positions]
-        return min((available_index for available_index in available_indexes),
-                   key=lambda index: abs(move_index - index))
-
-    while working_list:
-        # Pick which item to move
-        move_index = weighted_choice(
-            [(index, weight[2]) for index, weight in enumerate(working_list)])
-        move_item = working_list[move_index][0]
-        # Find the index where the item will be placed
-        if isinstance(working_list[move_index][1], str):
-            if working_list[move_index][1] == 'STAY':
-                # Place in the index closest to where the item appears already
-                target_position = closest_available(move_index)
+
+    def closest_available(requested_index, available_indices):
+        """The closest index to requested_index that is in the list of available_indices"""
+        # shuffles the available indexes to remove bias when finding the closest one
+        random.shuffle(available_indices)

```

Comment:
So, the purpose for this is to remove left or right bias in selecting the nearest value if there are 2 nearest values, depending on how the `min` function operates. 

For example, let's say I call `closest_available(3, [1,2,4,5])`, but skip the shuffle and the code directly goes to `min(available_indices, key=lambda index: abs(requested_index - index))`

In that case, there would be 2 indices that have a distance of 1 from the requested_index, so which one would be returned? It would depend on the evaluation of the `min` function, which is most likely iterating left -> right, so 2 would always be returned. Otherwise, 4 would always be returned, but either way it would be constant.

Adding the shuffle means that  `closest_available(3, [1,2,4,5])` will return either 2 or 4 randomly, as the min function it operating on shuffled indices.

That being said, `available`_indices should be somewhat random because of the shuffle I do on the original `weights` list, but I don't want to risk order arising due to the by-weight-sorting.